File: diagflow/rag/knowledge_base.py
"""
Knowledge base — closed-loop historical case reuse.

Two retrieval modes:
  1. MD5 fingerprint match (exact, zero LLM, Phase 1 fast-path)
  2. Hybrid semantic/keyword search (ChromaDB + BM25 + RRF, Phase 1 fallback)

Fingerprint is computed from the (component, error_pattern, version) tuple.
After each successful diagnosis, the case is auto-indexed — same issue next
time returns instantly. No manual case writing needed.
"""

from __future__ import annotations

import hashlib
import json
import logging
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from .embedder import Embedder
from .retriever import HybridRetriever
from .vector_store import VectorStore

logger = logging.getLogger(__name__)


class KnowledgeBase:
    """Closed-loop knowledge base with auto-indexing."""

    def __init__(
        self,
        vector_store: VectorStore | None = None,
        embedder: Embedder | None = None,
    ):
        self.vector_store = vector_store or VectorStore()
        if embedder is None:
            from diagflow.config import get_config
            cfg = get_config()
            embedder = Embedder(embedding_dim=cfg.vector_store.embedding_dim)
        self.embedder = embedder
        self.retriever = HybridRetriever(self.vector_store, self.embedder)
        # {md5_hash: KnownCase}
        self._fingerprints: dict[str, dict[str, Any]] = {}
        # Restore fingerprints from persisted ChromaDB metadata
        self._load_fingerprints()

    @property
    def _has_real_embeddings(self) -> bool:
        """True only when the embedder can produce semantically meaningful vectors.

        The hash fallback (no API key) is not semantically meaningful — but it
        is useless for recall and, if written, pollutes the vector store.
        """
        return self.embedder.api_key is not None

    # ------------------------------------------------------------------
    # Fingerprint persistence
    # ------------------------------------------------------------------

    def _load_fingerprints(self) -> None:
        """Rebuild in-memory fingerprint dict from ChromaDB metadata."""
        try:
            all_cases = self.vector_store.collection.get(
                include=["metadatas"]
            )
            if all_cases and all_cases.get("metadatas"):
                for i, meta in enumerate(all_cases["metadatas"]):
                    fp = meta.get("fingerprint", "")
                    if fp and fp not in self._fingerprints:
                        self._fingerprints[fp] = {
                            "case_id": all_cases["ids"][i] if all_cases.get("ids") else fp,
                            "component": meta.get("component", "unknown"),
                            "error_pattern": meta.get("error_pattern", "unknown"),
                            "version": meta.get("version", ""),
                            "root_cause": meta.get("root_cause", ""),
                            "suggestions": meta.get("suggestions", "").split(",") if isinstance(meta.get("suggestions"), str) else meta.get("suggestions", []),
                            "_hits": meta.get("_hits", 0),
                        }
                logger.info("Loaded %d fingerprints from ChromaDB", len(self._fingerprints))
        except Exception:
            logger.warning("Failed to load fingerprints from ChromaDB", exc_info=True)

    # ------------------------------------------------------------------
    # Fingerprint — fast-path (Phase 1, zero LLM)
    # ------------------------------------------------------------------

    @staticmethod
    def make_fingerprint(component: str, error_pattern: str, version: str = "") -> str:
        """MD5 of (component, error_pattern, version) — deterministic key."""
        raw = f"{component}:{error_pattern}:{version}"
        return hashlib.md5(raw.encode()).hexdigest()[:16]

    def fingerprint_match(
        self, component: str, error_pattern: str, version: str = ""
    ) -> dict | None:
        """Phase 1 fast-path: exact MD5 match. Returns cached case or None.

        Skips disputed cases (marked via mark_incorrect).
        """
        fp = self.make_fingerprint(component, error_pattern, version)
        match = self._fingerprints.get(fp)
        if match and not match.get("_disputed"):
            match["_hits"] = match.get("_hits", 0) + 1
            return match
        return None

    # ------------------------------------------------------------------
    # Indexing — auto-inserted after each successful diagnosis
    # ------------------------------------------------------------------

    def add_case(
        self,
        component: str,
        error_pattern: str,
        version: str,
        root_cause: str,
        suggestions: list[str],
        evidence_summary: str = "",
    ) -> str:
        """Index a case (called automatically by engine Phase 5)."""
        fp = self.make_fingerprint(component, error_pattern, version)
        case_id = f"{component}_{error_pattern}_{fp[:8]}"

        # Existing case? Update hit count and merge suggestions.
        existing = self._fingerprints.get(fp)
        if existing:
            existing["_hits"] = existing.get("_hits", 0) + 1
            existing["suggestions"] = list(
                set(existing.get("suggestions", []) + suggestions)
            )
            existing["last_seen"] = datetime.now(timezone.utc).isoformat()
            # Overwrite root_cause if we have a better one (not from empty restore)
            if root_cause and not existing.get("root_cause"):
                existing["root_cause"] = root_cause
            return fp

        # New case.
        case_text = (
            f"Component: {component} v{version}\n"
            f"Error pattern: {error_pattern}\n"
            f"Root cause: {root_cause}\n"
            f"Suggestions:\n"
            + "\n".join(f"- {s}" for s in suggestions)
        )

        embedding = self.embedder.embed(case_text)
        if self._has_real_embeddings:
            self.vector_store.add_case(
                case_id=case_id,
                text=case_text,
                metadata={
                    "component": component,
                    "error_pattern": error_pattern,
                    "version": version,
                    "fingerprint": fp,
                    "root_cause": root_cause,
                    "suggestions": ",".join(suggestions) if suggestions else "",
                    "created": datetime.now(timezone.utc).isoformat(),
                },
                embedding=embedding,
            )

        self._fingerprints[fp] = {
            "case_id": case_id,
            "component": component,
            "error_pattern": error_pattern,
            "version": version,
            "root_cause": root_cause,
            "suggestions": suggestions,
            "_hits": 1,
            "created": datetime.now(timezone.utc).isoformat(),
        }

        # Rebuild BM25 so new case is searchable.
        self._rebuild_bm25()
        return fp

    def add_case_from_text(self, text: str, metadata: dict[str, Any] | None = None) -> str | None:
        """Convenience: add from raw text, auto-extract error pattern.

        Called by engine Phase 5 after _kb_index().
        """
        meta = metadata or {}
        component = meta.get("component", "unknown")
        version = meta.get("version", "")
        # Extract error pattern from text — simple keyword extraction.
        error_pattern = self._extract_error_pattern(text)
        root_cause = text.split("\n")[0][:200] if text else "unknown"
        suggestions = (
            [l[2:] for l in text.split("\n") if l.startswith("- ")]
            if text else []
        )
        return self.add_case(component, error_pattern, version, root_cause, suggestions)

    # ------------------------------------------------------------------
    # Search — Phase 1 fallback (semantic + keyword)
    # ------------------------------------------------------------------

    def search(self, query: str, n: int = 5) -> list[dict[str, Any]]:
        """Hybrid search across all indexed cases."""
        return self.retriever.search(query, n_results=n)

    # ------------------------------------------------------------------
    # Bulk loading — seed cases from Markdown files
    # ------------------------------------------------------------------

    def add_case_from_md(self, md_path: str) -> None:
        """Index a Markdown case file (for seed data)."""
        path = Path(md_path)
        if not path.exists():
            raise FileNotFoundError(f"Case file not found: {md_path}")

        text = path.read_text(encoding="utf-8")
        meta = self._parse_frontmatter(text)
        body = self._strip_frontmatter(text)

        case_id = path.stem
        component = meta.get("component", "unknown")
        error_pattern = meta.get("error_pattern", "unknown")
        version = meta.get("version", "")
        # Extract root cause: prefer frontmatter, then first non-heading line
        root_cause = meta.get("root_cause", "")
        if not root_cause:
            for line in body.split("\n"):
                stripped = line.strip()
                if stripped and not stripped.startswith("#"):
                    root_cause = stripped[:200]
                    break
            if not root_cause:
                root_cause = "unknown"

        # Fingerprint COMPUTED, not read from file.
        fp = self.make_fingerprint(component, error_pattern, version)

        embedding = self.embedder.embed(body)
        if self._has_real_embeddings:
            self.vector_store.add_case(
                case_id=case_id,
                text=body,
                metadata={**meta, "fingerprint": fp},
                embedding=embedding,
            )
        self._fingerprints[fp] = {
            "case_id": case_id,
            "component": component,
            "error_pattern": error_pattern,
            "version": version,
            "root_cause": root_cause,
            "suggestions": meta.get("suggestions", "").split(","),
            "_hits": 0,
        }

    def load_cases_from_dir(self, dir_path: str) -> int:
        """Load all .md case files from a directory."""
        count = 0
        for f in Path(dir_path).glob("*.md"):
            try:
                self.add_case_from_md(str(f))
                count += 1
            except Exception as exc:
                logger.warning("Failed to load case %s: %s", f, exc)
        self._rebuild_bm25()
        return count

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    @staticmethod
    def _extract_error_pattern(text: str) -> str:
        """Extract key error keywords from text."""
        from diagflow.config import get_config
        cfg = get_config()
        keywords = cfg.rag.error_keywords
        found = [k for k in keywords if k in text]
        return found[0] if found else text[:50].strip()

    @staticmethod
    def _parse_frontmatter(text: str) -> dict[str, Any]:
        metadata: dict[str, Any] = {}
        if text.startswith("---"):
            parts = text.split("---", 2)
            if len(parts) >= 3:
                for line in parts[1].strip().split("\n"):
                    if ":" in line:
                        k, v = line.split(":", 1)
                        metadata[k.strip()] = v.strip()
        return metadata

    @staticmethod
    def _strip_frontmatter(text: str) -> str:
        if text.startswith("---"):
            parts = text.split("---", 2)
            if len(parts) >= 3:
                return parts[2].strip()
        return text

    def _rebuild_bm25(self) -> None:
        all_cases = self.vector_store.collection.get(
            include=["documents", "metadatas"]
        )
        if all_cases and all_cases.get("documents"):
            self.retriever.build_index(
                documents=all_cases["documents"],
                ids=all_cases["ids"],
                metadatas=all_cases["metadatas"],
            )

    def stats(self) -> dict[str, Any]:
        return {
            "fingerprint_cases": len(self._fingerprints),
            "chroma_docs": self.vector_store.count(),
        }

    def mark_incorrect(self, fp: str) -> None:
        """Mark a fingerprint as disputed — prevents future fast-path hits."""
        if fp in self._fingerprints:
            self._fingerprints[fp]["_disputed"] = True
            logger.info("Marked fingerprint as disputed: %s", fp)

    def on_feedback_incorrect(self, component: str, error_pattern: str, version: str = "",
                              comment: str = "") -> str | None:
        """User marked a diagnosis as incorrect → dispute the case in KB.

        Returns the disputed fingerprint, or None if no match found.
        """
        fp = self.make_fingerprint(component, error_pattern, version)
        self.mark_incorrect(fp)
        if comment:
            self._fingerprints[fp]["_dispute_comment"] = comment
        return fp

File: internal/rag/knowledgebase/knowledgebase.go
// Package knowledgebase mirrors diagflow/rag/knowledge_base.py — closed-loop
// historical case reuse with MD5 fingerprint fast-path and hybrid search.
package knowledgebase

import (
	"crypto/md5"
	"encoding/hex"
	"strings"
	"sync"

	"github.com/Admaing/diagflow/internal/config"
	"github.com/Admaing/diagflow/internal/rag/embedder"
	"github.com/Admaing/diagflow/internal/rag/retriever"
	"github.com/Admaing/diagflow/internal/rag/vectorstore"
)

// KnownCase is a cached fingerprint entry.
type KnownCase struct {
	CaseID       string
	Component    string
	ErrorPattern string
	Version      string
	RootCause    string
	Suggestions  []string
	Hits         int
	Disputed     bool
}

// KnowledgeBase manages fingerprints + hybrid retrieval.
type KnowledgeBase struct {
	mu           sync.Mutex
	Store        vectorstore.VectorStore
	Embedder     *embedder.Embedder
	Retriever    *retriever.HybridRetriever
	fingerprints map[string]*KnownCase
}

// New builds a KnowledgeBase.
func New(store vectorstore.VectorStore) *KnowledgeBase {
	cfg := config.Get()
	emb := embedder.New("text-embedding-3-small", cfg.OpenAIAPIKey, cfg.VectorStore.EmbeddingDim)
	kb := &KnowledgeBase{
		Store:        store,
		Embedder:     emb,
		fingerprints: make(map[string]*KnownCase),
	}
	kb.Retriever = retriever.New(store, emb, 60)
	kb.loadFingerprints()
	return kb
}

// MakeFingerprint computes MD5(component:error:version)[:16].
func MakeFingerprint(component, errorPattern, version string) string {
	raw := component + ":" + errorPattern + ":" + version
	sum := md5.Sum([]byte(raw))
	return hex.EncodeToString(sum[:])[:16]
}

// HasRealEmbeddings reports whether the KB can produce semantically meaningful
// vectors (i.e. an embedding API key is present).
func (kb *KnowledgeBase) HasRealEmbeddings() bool {
	return kb.Embedder.HasRealEmbeddings()
}

func (kb *KnowledgeBase) loadFingerprints() {
	all := kb.Store.Collection().Get([]string{"metadatas"})
	if all.Metadatas == nil {
		return
	}
	for i, meta := range all.Metadatas {
		fp, _ := meta["fingerprint"].(string)
		if fp == "" {
			continue
		}
		if _, ok := kb.fingerprints[fp]; ok {
			continue
		}
		caseID := ""
		if i < len(all.IDs) {
			caseID = all.IDs[i]
		}
		kb.fingerprints[fp] = &KnownCase{
			CaseID:       caseID,
			Component:    strOr(meta["component"], "unknown"),
			ErrorPattern: strOr(meta["error_pattern"], "unknown"),
			Version:      strOr(meta["version"], ""),
			RootCause:    strOr(meta["root_cause"], ""),
			Suggestions:  toSuggestions(meta["suggestions"]),
		}
	}
}

// FingerprintMatch returns a non-disputed hit, or nil.
func (kb *KnowledgeBase) FingerprintMatch(component, errorPattern, version string) map[string]any {
	kb.mu.Lock()
	defer kb.mu.Unlock()
	fp := MakeFingerprint(component, errorPattern, version)
	if m, ok := kb.fingerprints[fp]; ok && !m.Disputed {
		m.Hits++
		return map[string]any{
			"root_cause":    m.RootCause,
			"suggestions":   m.Suggestions,
			"component":     m.Component,
			"error_pattern": m.ErrorPattern,
			"version":       m.Version,
		}
	}
	return nil
}

// AddCase indexes a case (called after a successful diagnosis).
func (kb *KnowledgeBase) AddCase(component, errorPattern, version, rootCause string, suggestions []string) string {
	fp := MakeFingerprint(component, errorPattern, version)
	caseID := component + "_" + errorPattern + "_" + fp[:8]

	kb.mu.Lock()
	if existing, ok := kb.fingerprints[fp]; ok {
		existing.Hits++
		existing.Suggestions = mergeStr(existing.Suggestions, suggestions)
		if rootCause != "" && existing.RootCause == "" {
			existing.RootCause = rootCause
		}
		kb.mu.Unlock()
		return fp
	}
	kb.mu.Unlock()

	var b strings.Builder
	b.WriteString("Component: " + component + " v" + version + "\n")
	b.WriteString("Error pattern: " + errorPattern + "\n")
	b.WriteString("Root cause: " + rootCause + "\n")
	b.WriteString("Suggestions:\n")
	for _, s := range suggestions {
		b.WriteString("- " + s + "\n")
	}
	caseText := b.String()

	if kb.Embedder.HasRealEmbeddings() {
		embedding := kb.Embedder.Embed(caseText)
		kb.Store.AddCase(caseID, caseText, map[string]any{
			"component":     component,
			"error_pattern": errorPattern,
			"version":       version,
			"fingerprint":   fp,
			"root_cause":    rootCause,
			"suggestions":   strings.Join(suggestions, ","),
		}, embedding)
	}

	kb.mu.Lock()
	kb.fingerprints[fp] = &KnownCase{
		CaseID:       caseID,
		Component:    component,
		ErrorPattern: errorPattern,
		Version:      version,
		RootCause:    rootCause,
		Suggestions:  suggestions,
		Hits:         1,
	}
	kb.mu.Unlock()

	kb.RebuildBM25()
	return fp
}

// Search runs hybrid search across all indexed cases.
func (kb *KnowledgeBase) Search(query string, n int) []vectorstore.SearchResult {
	return kb.Retriever.Search(query, n, 0.6, 0.4)
}

// RebuildBM25 rebuilds the retriever index from the store.
func (kb *KnowledgeBase) RebuildBM25() {
	kb.Retriever.RebuildFromStore()
}

// MarkIncorrect disputes a fingerprint, preventing future fast-path hits.
func (kb *KnowledgeBase) MarkIncorrect(fp string) {
	kb.mu.Lock()
	defer kb.mu.Unlock()
	if m, ok := kb.fingerprints[fp]; ok {
		m.Disputed = true
	}
}

// AddInvestigationIntent stores an investigation intent (排查思路) linked to its
// root cause, so a future similar intent can semantically match and short-circuit.
// Only called after the user confirms a diagnosis was correct.
func (kb *KnowledgeBase) AddInvestigationIntent(intentDesc, rootCause, component, version string) {
	if !kb.Embedder.HasRealEmbeddings() {
		return // semantic retrieval requires a real embedding backend
	}
	caseID := "intent_" + MakeFingerprint(component, intentDesc, version)[:12]
	text := "排查思路: " + intentDesc + "\n根因: " + rootCause + "\n组件: " + component
	kb.Store.AddCase(caseID, text, map[string]any{
		"component":  component,
		"root_cause": rootCause,
		"version":    version,
	}, kb.Embedder.Embed(text))
	kb.RebuildBM25()
}

// Stats returns a summary.
func (kb *KnowledgeBase) Stats() map[string]any {
	kb.mu.Lock()
	n := len(kb.fingerprints)
	kb.mu.Unlock()
	return map[string]any{
		"fingerprint_cases": n,
		"chroma_docs":       kb.Store.Count(),
	}
}

func strOr(v any, def string) string {
	if s, ok := v.(string); ok {
		return s
	}
	return def
}

func toSuggestions(v any) []string {
	switch t := v.(type) {
	case string:
		if t == "" {
			return nil
		}
		return strings.Split(t, ",")
	case []any:
		out := make([]string, 0, len(t))
		for _, item := range t {
			if s, ok := item.(string); ok {
				out = append(out, s)
			}
		}
		return out
	case []string:
		return t
	default:
		return nil
	}
}

func mergeStr(a, b []string) []string {
	seen := make(map[string]bool)
	var out []string
	for _, list := range [][]string{a, b} {
		for _, s := range list {
			if !seen[s] {
				seen[s] = true
				out = append(out, s)
			}
		}
	}
	return out
}
